Batteries/Data/Vector/Basic.lean

```diff
@@ -0,0 +1,493 @@
+def drop (i : Nat) (v : Vector α n) : Vector α (n - i) :=
+  have : min n n - i = n - i := by
+    simp only [Nat.min_self]
+  this.symm ▸ (extract v i n)
+
+/-- Take `i` elements from a vector of length `n`; we can have `i > n`. -/
+def take (i : Nat) (v : Vector α n) : Vector α (min i n) :=
+  extract v 0 i
+
+/-- Vector of length `n` from a function on `Fin n`. -/
+def ofFn (f : Fin n → α) : Vector α n := ⟨Array.ofFn f, by {simp only [size_ofFn]}⟩
+
+/-- Create a vector from another with a provably equal length. -/
+protected def cast {n m : Nat} (h : n = m) : Vector α n → Vector α m
+  | ⟨x, p⟩ => ⟨x, h ▸ p⟩
+
+
+-- Swap functions
+/--
+Swaps two entries in an array, or panics if either index is out of bounds.
+
+This will perform the update destructively provided that `v` has a reference
+count of 1 when called.
+-/
+
+def swap (v : Vector α n) (i j : Fin n) : Vector α n :=
+  ⟨Array.swap v.toArray (v.size_eq.symm ▸ i) (v.size_eq.symm ▸ j), proof⟩
+  where
+    proof := by
+      simp only [size_swap]
+      exact v.size_eq
+      done
+
+/--
+Swaps two entries in an vector, or panics if either index is out of bounds.
+
+This will perform the update destructively provided that `v` has a reference count of 1 when called.
+-/
+def swap! (v : Vector α n) (i j : Nat) : Vector α n :=
+  ⟨Array.swap! v.toArray i j, by {simp only [size_swap!]; exact v.size_eq}⟩
+
+/--
+Swaps the entry with index `i : Fin n` in the vector for a new entry.
+The old entry is returned with the modified vector.
+
+This will perform the update destructively provided that `v` has a reference count of 1 when called.
+-/
+def swapAt (v : Vector α n) (i : Fin n) (x : α) : α × Vector α n:=
+  let e := Vector.get v i
+  let v := Vector.set v i x
+  (e, v)
+
+
+instance [Inhabited α] : Inhabited (Vector α n) where
+  default := mkVector n default
+
+/--
+`swapAt! v i x` Swaps out the entry at index `i : Nat` in the vector for `x`, if the index is valid.
+Otherwise it panics
+The old entry is returned with the modified vector.
+
+This will perform the update destructively provided that `v` has a reference count of 1 when called.
+-/
+def swapAt! (v : Vector α n) (i : Nat) (x : α) : α × Vector α n :=
+  if h : i < n then
+    swapAt v ⟨i, h⟩ x
+  else
+    have : Inhabited α := ⟨x⟩
+    panic! ("index " ++ toString i ++ " out of bounds")
+
+/--`range n` Returns a vector `#v[1,2,3,...,n-1]`-/
+def range (n : Nat) : Vector Nat n :=
+  ⟨Array.range n, by {simp}⟩
+
+/--
+Returns a vector of size `1` with a single element `v`
+-/
+def singleton (v : α) : Vector α 1 :=
+  mkVector 1 v
+
+/--
+Vector lookup function that takes an index `i` of type `USize`
+-/
+def uget (v : Vector α n) (i : USize) (h : i.toNat < n) : α :=
+  Array.uget v.toArray i (v.size_eq.symm ▸ h)
+
+
+/--
+borrowed from array API:
+auxiliary declaration used in the equation compiler when pattern matching array literals.
+-/
+abbrev getLit (v : Vector α n) (i : Nat) (h₂ : i < n) : α :=
+  Vector.get v ⟨i, h₂⟩
```

Comment:
I don't think this does anything for vector. Someone who knows how `Array.getLit` interacts with the equation compiler could help here. Otherwise, delete.

Reply:
I was correct. This code does nothing for vectors. Delete it.

Reply:
done